src/sgame/botlib/bot_load.cpp

```diff
@@ -79,101 +103,231 @@ void BotSaveOffMeshConnections( NavData_t *nav )
+
+		trap_FS_Read( con.verts, sizeof( float ) * 6 * conCount, f );
+		SwapArray( con.verts, conCount * 6 );
+
+		trap_FS_Read( con.rad, sizeof( float ) * conCount, f );
+		SwapArray( con.rad, conCount );
+
+		trap_FS_Read( con.flags, sizeof( unsigned short ) * conCount, f );
+		SwapArray( con.flags, conCount );
+
+		trap_FS_Read( con.areas, sizeof( unsigned char ) * conCount, f );
+		trap_FS_Read( con.dirs, sizeof( unsigned char ) * conCount, f );
+
+		// userids is loaded but unused.
+		trap_FS_Read( con.userids, sizeof( unsigned int ) * conCount, f );
+		SwapArray( con.userids, conCount );
+
+		Log::Debug( "Loaded %d connections from navcon file %s", conCount, filePath );
 		trap_FS_FCloseFile( f );
+
+		BotPrintOffMeshConnections( con );
 		return;
 	}
 
-	int conCount = header.numConnections;
+	// Seek to the beginning of the file.
+	trap_FS_Seek( f, 0, fsOrigin_t::FS_SEEK_SET );
+
+	// New text format.
+	char *navcon = (char*) BG_Alloc( len + 1 );
+	char *navconHead = navcon;
 
-	if ( conCount > con.MAX_CON )
+	trap_FS_Read( navcon, len, f );
+	trap_FS_FCloseFile( f );
+	navcon[ len ] = '\0';
+
+	// Do not read numbers from the next line with sscanf;
+	char *endline = strchr( navcon, '\n' );
+	if ( endline != nullptr )
 	{
-		trap_FS_FCloseFile( f );
+		*endline = '\0';
+	}
+
+	if ( strncmp( navcon, NAVCON_HEADER_PREFIX, strlen( NAVCON_HEADER_PREFIX ) ) != 0 )
+	{
+		Log::Warn( "Unknown format for navcon file %s", filePath );
 		return;
 	}
 
-	con.offMeshConCount = conCount;
+	// The dummy is used to detect when there are more fields than needed.
+	char dummy;
 
-	trap_FS_Read( con.verts, sizeof( float ) * 6 * conCount, f );
-	SwapArray( con.verts, conCount * 6 );
+	int scanned = sscanf( navcon, va( "%s %%d %%c", NAVCON_HEADER_PREFIX ), &version, &dummy );
 
-	trap_FS_Read( con.rad, sizeof( float ) * conCount, f );
-	SwapArray( con.rad, conCount );
+	if ( scanned != 1 )
+	{
+		if ( scanned > 1 )
+		{
+			Log::Warn( "Malformed navcon header, extra field after format version in navcon file %s: %s", filePath, navcon );
+			return;
+		}
+
+		Log::Warn( "Missing format version for navcon file %s", filePath );
+		return;
+	}
+
+	if ( version != NAVCON_VERSION )
+	{
+		if ( version == 2 )
+		{
+			Log::Warn("Format version %d should not be a text format for navcon file %s", version, filePath );
+			return;
+		}
 
-	trap_FS_Read( con.flags, sizeof( unsigned short ) * conCount, f );
-	SwapArray( con.flags, conCount );
+		Log::Warn("Unknown format version %d for navcon file %s", version, filePath );
+		return;
+	}
 
-	trap_FS_Read( con.areas, sizeof( unsigned char ) * conCount, f );
-	trap_FS_Read( con.dirs, sizeof( unsigned char ) * conCount, f );
+	Log::Debug("Loading format version %d navcon file %s", version, filePath );
+	navcon += strlen( va( "%s %d\n", NAVCON_HEADER_PREFIX, version) );
 
-	trap_FS_Read( con.userids, sizeof( unsigned int ) * conCount, f );
-	SwapArray( con.userids, conCount );
+	char line[ MAX_STRING_CHARS ];
```

Comment:
May I suggest a more economical way to iterate lines:

```
for ( int pos = 0; pos < len; )
{
   int lineLength = strcspn( navcon + pos, "\n" );
   navcon[ pos + lineLength ] ='\0';
   
  sscanf( navcon + pos, ...);
  // etc.

   pos += lineLength + 1;
}

Reply:
Right, if that works we may also backport it to the layout parser (I initially copy pasted the layout parser).

Reply:
Done,